mpact_api/mpact/views.py

```diff
@@ -1,3 +1,78 @@
-from django.shortcuts import render
+import os
 
-# Create your views here.
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from .models import ChatData, UserChat, UserData
+from .serializers import UserDataSerializer
+from .services import save_chatdata, save_userchat, save_userdata, set_webhook
+
+
+class Webhook(APIView):
+    """
+    This is only called once for the webhook setup.
+    """
+
+    def get(self, request):
+        webhook = set_webhook()
+        if webhook:
+            return Response("webhook setup ok")
+        else:
+            return Response("webhook setup failed")
+
+
+class ListenMessages(APIView):
+    """
+    Telegram webhook makes a post request.
+    """
+
+    def post(self, request):
+        message = request.data["message"]
+        chat_data = message.get("chat")
```

Comment:
I see you are using `message.get()`, and later, `chat_data.get()` and `from_data.get()`. While this will allow this code not to error out if `message` does not look the way you expect, you will either end up with errors later, because `None` doesn't have a `get()` method, or you will end up with unexpected/missing data in your database.

It might be better to validate that `message` looks the way you expect, and has the data you need, as soon as possible.

A great library for doing this is [pydantic](https://pypi.org/project/pydantic/) but you could also use [schema](https://pypi.org/project/schema/) or even [jsonobject](https://pypi.org/project/jsonobject/). (Don't use "jsonobject". :upside_down_face: Dimagi wrote it, but "schema" is better, and "pydantic" is awesome.)

Reply:
Actually, from the webhook, we will always get the "message", "from" and "chat" in the request data whereas "new_chat_participant" will only be present when a new participant is added, therefore it is being extracted in If condition. That is the reason I didn't think of such validations.  

But later we may need it, so it is better to add it from starting only. Out of curiosity, I wanted to ask you that anyhow we are validating the data using serializers before storing in the DB. So do we really need pydantic? or How pydantic will be different from django serializers?

Reply:
pydantic or schema might make it easier to validate `request.data["message"]` on the very first line of the `post()` method ... but using Django serializers is a great idea! Let's go with that. It means we don't have to introduce any new dependencies.